[Sema] Fix `try`/`await` handling for `if`/`switch` expressions

hamishknight opened the pull request:
Previously we allowed an attached `try` or `await` on an `if`/`switch` expression to cover the branches, which does not match what was proposed, and is especially harmful for multi-statement cases. Fix the effect handling logic such that we reset effect coverage for `if`/`switch` expressions similar to closures, but continue to maintain the state needed for rethrows checking. This means that we'll now consistently diagnose `try` and `await` on `if`/`switch` expressions as invalid.

rdar://116066748

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

xedin commented on test/expr/unary/if_expr.swift:
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-typecheck-verify-swift -disable-availability-checking
+// RUN: %target-typecheck-verify-swift -disable-availability-checking -enable-experimental-concurrency
```
Should this also use `// REQUIRES: concurrency` and `// REQUIRES: asserts`?

hamishknight replied:
Ah yeah, good catch! Probably ought to separate out the tests that need this

hamishknight committed:
[Sema] Fix `try`/`await` handling for `if`/`switch` expressions

Previously we allowed an attached `try` or `await`
on an `if`/`switch` expression to cover the branches,
which does not match what was proposed, and is
especially harmful for multi-statement cases. Fix
the effect handling logic such that we reset
effect coverage for `if`/`switch` expressions
similar to closures, but continue to maintain the
state needed for rethrows checking.

rdar://116066748

hamishknight commented:
@swift-ci please test
